Further transliteration of typing rules

jalglave opened the pull request:
@HadrienRenaud please could you have a look?
thanks in advance,
Jade

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -722,124 +722,261 @@ of the following applies:
     {4 Rule}
+    All of the following applies:
+    - One of the following applies:
+      - [s] is a return statement with some value and no return type;
+      - [s] is a return statement with no value and some return type;
```
A generalization of my previous comment:
```suggestion
      - All of the following applies:
        - [s] is a return statement with some value;
        - the enclosing function does not have a return type;
      - All of the following applies:
        - [s] is a return statement with no value;
        - the enclosing function does have a return type;
```

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -722,124 +722,261 @@ of the following applies:
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a return statement with some value [e] and some return type [t];
```
```suggestion
    - [s] is a return statement with some value [e];
    - the enclosing function has a return type [t];
```

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -722,124 +722,261 @@ of the following applies:
+    - [t2,e2'] is the result of annotating [e2] in [env];
+    - [t1] has the structure of an integer type;
+    - [t2] has the structure of an integer type;
+    - [ty] is ;
+    - [s''] is the result of annotating [s'] in [env];
```
```suggestion
    - [ty] is the integer type [integer {e1' .. e2'}];
    - [env1] is [env] with the identifier [id] locally declared as immutable of type [ty];
    - [s''] is the result of annotating [s'] in [env1];
```

Note that the type `ty` is actually more complicated as specified by rule ZSND (paragraph 6.9 of the LRM). We might want to specify this here.

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -722,124 +722,261 @@ of the following applies:
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a declaration with [ldk], [ldi] and an optional expression [e_opt];
+    - [e_opt] gives some expression [e];
```
I think this is equivalent and more explicit (we have a degree less of complexity)
```suggestion
    - [s] is a declaration with [ldk], [ldi] and an expression [e];
```

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -722,124 +722,261 @@ of the following applies:
+    - [s] is a declaration with [ldk], [ldi] and an optional expression [e_opt];
+    - [e_opt] gives some expression [e];
+    - [t_e,e'] is the result of annotating [e] in [env];
+    - [env', ldi'] is the result of declaring the local constants of [ldi] in [env];
```
```suggestion
    - [env', ldi'] is the result of declaring the local identifiers of [ldi] in [env];
```

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -722,124 +722,261 @@ of the following applies:
+    - [t_e,e'] is the result of annotating [e] in [env];
+    - [env', ldi'] is the result of declaring the local constants of [ldi] in [env];
+    - [new_s] is a declaration with [ldk], [ldi'] and an expression [e'];
+    - [new_env] is [env].
```
```suggestion
    - [new_env] is [env'].
```

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -722,124 +722,261 @@ of the following applies:
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a declaration statement with [ldk], [ldi] and an optional expression [e_opt];
+    - [e_opt] gives no expression;
```
```suggestion
    - [s] is a declaration statement with [ldk], [ldi] and no initial expression;
```

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -722,124 +722,261 @@ of the following applies:
+    - [e_opt] gives no expression;
+    - [env', s'] is the result of annotating uninitialised local declarations [ldi] in [env];
+    - [new_s] is [s'];
+    - [new_env] is [env].
```
```suggestion
    - [new_env] is [env'].
```

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -722,124 +722,261 @@ of the following applies:
+    - [s] is a throw statement with expression [e];
+    - [t_e,e'] is the result of annotating [e] in [env];
+    - [t_e] has the structure of an exception type;
+    - [new_s] is a throw statement with expression [e'] and type [t_e];
```
This is actually not syntax, so we might want to specify somewhere that we are adding some typing information in the AST.
Maybe `AST.mli` deserves an intro on that point?

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -722,124 +722,261 @@ of the following applies:
 
     {4 Example}
 
   {3 TypingRule.SDebug}
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a debug statement with expression [e];
+    - [_t_e, e'] is the result of annotating [e] in [env];
+    - [new_s] is a debug statement with expression [e'];
+    - [new_env] is [env].
```
I do not think this should be in the spec. This is unspecified, and a very-asllib only syntax and code.

Probably once we fix the type-checking of the procedure `print`, all the `debug` statements could be converted in the ASL into `print` statements.

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -981,54 +981,91 @@ one of the following applies:
+    All of the following applies:
+    - [slices] gives [(factor, pre_length)];
+    - [pre_offset] is [factor * pre_length];
+    - [offset, length] is the result of applying TypingRule.SliceLength to [(pre_offset, length)].
```
```suggestion
    - [offset, length] is the result of applying TypingRule.SliceLength to [(pre_offset, pre_length)].
```

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -1041,30 +1078,92 @@ Formally, one of the following applies:
+    All of the following applies:
+    - [p] is the pattern which matches anything greater than or equal to an expression [e];
+    - [t_e, e'] is the result of annotating [e] in [env]; 
+    - [e'] is pure;
```
FIY, the LRM asks for a stronger check: that `e` (or `e'` similarly) is a compile-time constant.

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -1041,30 +1078,92 @@ Formally, one of the following applies:
+    - [t] has the structure of an integer;
+    - [t_e1] has the structure of an integer;
+    - [t_e2] has the structure of an integer;
+    (*no need to check that e1' and e2' are pure here?*)
```
Here we would need to check that `e1` and `e2` are compile-time constants..


I created HadrienRenaud/herdtools7#39 (and HadrienRenaud/herdtools7#40 which is related).

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -1041,30 +1078,92 @@ Formally, one of the following applies:
+    - [p] is the pattern which matches a mask [m];
+    - [t] has the structure of a bitvector type;
+    - [n] is the length of mask [m];
+    - [t_m] is the bitvectore type of width [n];
```
```suggestion
    - [t_m] is the bitvector type of width [n];
```

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -1041,30 +1078,92 @@ Formally, one of the following applies:
+    - [t] has the structure of a bitvector type;
+    - [n] is the length of mask [m];
+    - [t_m] is the bitvectore type of width [n];
+    - [t_m] type-satisfies [t];
```
```suggestion
    - [t] type-satisfies [t_m];
```

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -1041,30 +1078,92 @@ Formally, one of the following applies:
+    {4 Rule}
+    All of the following applies:
+    - [p] is the pattern which matches a tuple [li];
+    - [t_struct] is the type structure of [t];
+    - [t_struct] is a tuple [ts];
```
```suggestion
    - [t] has the structure of a tuple type [ts];
```

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -1041,30 +1078,92 @@ Formally, one of the following applies:
     {4 Rule}
+    All of the following applies:
+    - [p] is the pattern which matches a tuple [li];
+    - [t_struct] is the type structure of [t];
```
FYI this is probably not what we exactly want: we would like `p_struct` to be the anonymous type corresponding to `t` in `env`

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
+    From that follows that we can discard the environment at the end of an
+    enclosing block.
+    
     {4 Example}
```
```suggestion
    {4 Example}
    [@asl{
    func main () => integer
    begin
      if UNKNOWN: boolean then
         let i = 3;
         print (i);
      end
      let i = "Some text";
      print (i);
      return 0;
    end
    }]
```

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
-[annotate_catchers env catchers otherwise_opt s_m]
+  
+  {3 TypingRule.CatcherNone}
+  [annotate_catchers env return_type (name_opt, ty, stmt)] is [(name_opt, ty,
```
It feels strange to me that `return_type` is not hidden in the local `env`. I might refacto that soon.

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
+     - [ty] has the structure of an exception type;
+    - [name_opt] gives a name [name];
+    - [name] is not already declared in [env];
+    - [name] is annotated with [ty] in [env];
+    - [env'] is [env];
```
```suggestion
    - [env'] is [env] with [name] locally declared as immutable of type [ty];
```

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
     {4 Rule}
-    <description>
+    All of the following applies:
+    - [caller_arg_types, arg1] is the result of annotating [args] in [env];
```
```suggestion
    - [caller_arg_types, args1] is the result of annotating [args] in [env];
```

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
-    <description>
+    All of the following applies:
+    - [caller_arg_types, arg1] is the result of annotating [args] in [env];
+    - [extra_nargs, name1, callee_arg_types, ret_ty] is the result of trying to
+      find the name [name] with [caller_arg_types] in [env];
```
I find the following a bit more understandable, but I'm not sure it's that better:

```suggestion
    - [name] is bound in [env] to a function with argument types [callee_arg_types];
```

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
+    - [caller_arg_types, arg1] is the result of annotating [args] in [env];
+    - [extra_nargs, name1, callee_arg_types, ret_ty] is the result of trying to
+      find the name [name] with [caller_arg_types] in [env];
+    - [eqs1] is the list made of both [eqs] and [extra_nargs];
```
```suggestion
```

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
+    {4 Rule}
+    All of the following applies:
+    - [caller_arg_types, arg1] is the result of annotating [args] in [env];
+    - [extra_nargs, name1, callee_arg_types, ret_ty] is the result of trying to
+      find the name [name] with [caller_arg_types] in [env];
```
```suggestion
    - [name] is bound in [env] to a function with a unique name [name1] whose argument types [callee_arg_types] type-clash [caller_arg_types] and whose return type is [ret_ty];
```

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
+    - [eqs1] is the list made of both [eqs] and [extra_nargs];
+    - [eqs2] is ;
+    - [call_type] is a either a function or a getter type;
+    - [ret_ty] is a type [ty];
```
```suggestion
```

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
+    - [extra_nargs, name1, callee_arg_types, ret_ty] is the result of trying to
+      find the name [name] with [caller_arg_types] in [env];
+    - [eqs1] is the list made of both [eqs] and [extra_nargs];
+    - [eqs2] is ;
```
```suggestion
    - [eqs2] is [eqs1] with added all the type equations from the type-clash of [caller_arg_types] and [callee_arg_types];
```

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
+  - [env2] is ;
+  - [env3] is [env2] modified to add explicit parameters;
+  - [env4] is [env3] modified to resolve dependently typed identifiers in the arguments;
+  - [env5] is [env4] modified to resolve dependently typed identifiers in the result type;
```
Are we still doing that? That shouldn't be the case.

HadrienRenaud replied:
Raised a new issue: https://github.com/HadrienRenaud/herdtools7/issues/41

HadrienRenaud commented on asllib/ASLTypingReference.mld:
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
   {4 Rule}
+  All of the following applies:
+  - [env1] is [env] modified to have an empty local environment;
+  - [env2] is ;
```
```suggestion
  - [env2] is [env1] with every formal argument declared as immutable with its type;
```

HadrienRenaud reviewed (commented):
Hi @jalglave,

I will have to come back to some points on this review, as this points to a certain number of TODOs on my side.
There are certain comments that apply to other very similar rules.

Feel free to argue me out of those suggestions!

jalglave commented:
thanks @HadrienRenaud I think I've addressed all your comments - please could you double-check, and if I've not missed anything, please could you merge?

HadrienRenaud commented:
Hi @jalglave, I have marked as resolved the comments I think you've addressed. There are a few left, can you double check that you do not want them?
For example, this one looks like you've missed something: https://github.com/herd/herdtools7/pull/671/files#r1334979702

Otherwise, I'm happy to run tests, squash and merge when you want.
